sw/device/tests/i2c_host_irq_test.c

```diff
@@ -0,0 +1,298 @@
+static dif_rv_plic_t plic;
+
+// Hold the test result.
+static volatile status_t isr_result;
+// Used to sync the irs and the main thread.
+static volatile dif_i2c_irq_t irq_fired;
+
+/**
+ * Provides external IRQ handling for this test.
+ *
+ * This function overrides the default OTTF external ISR.
+ *
+ * For each IRQ, it performs the following:
+ * 1. Claims the IRQ fired (finds PLIC IRQ index).
+ * 2. Checks that the index belongs to the expected peripheral.
+ * 3. Checks that only the correct / expected IRQ is triggered.
+ * 4. Clears the IRQ at the peripheral.
+ * 5. Completes the IRQ service at PLIC.
+ */
+static status_t external_isr(void) {
+  dif_rv_plic_irq_id_t plic_irq_id;
+  TRY(dif_rv_plic_irq_claim(&plic, kHart, &plic_irq_id));
+
+  top_earlgrey_plic_peripheral_t peripheral = (top_earlgrey_plic_peripheral_t)
+      top_earlgrey_plic_interrupt_for_peripheral[plic_irq_id];
+  TRY_CHECK(peripheral == kTopEarlgreyPlicPeripheralI2c2,
+            "IRQ from incorrect peripheral: exp = %d(i2c2), found = %d",
+            kTopEarlgreyPlicPeripheralI2c2, peripheral);
+
+  irq_fired =
+      (dif_i2c_irq_t)(plic_irq_id - (dif_rv_plic_irq_id_t)
+                                        kTopEarlgreyPlicIrqIdI2c2FmtThreshold);
+
+  LOG_INFO("%s: plic:%d, i2c:%d", __func__, plic_irq_id, irq_fired);
+  TRY(dif_i2c_irq_acknowledge(&i2c, irq_fired));
+
+  // Complete the IRQ at PLIC.
+  TRY(dif_rv_plic_irq_complete(&plic, kHart, plic_irq_id));
+  return OK_STATUS();
+}
+
+void ottf_external_isr(void) {
+  LOG_INFO("%s", __func__);
+  status_t tmp = external_isr();
+  if (status_ok(isr_result)) {
+    isr_result = tmp;
+  }
+}
+
+// We make this a macro to show the caller line number when printing a error.
+#define CHECK_IRQ_EQ(irq)                                                \
+  do {                                                                   \
+    irq_global_ctrl(false);                                              \
+    if (irq_fired == UINT32_MAX) {                                       \
+      wait_for_interrupt();                                              \
+      irq_global_ctrl(true);                                             \
+    }                                                                    \
+    TRY_CHECK(irq_fired == irq, "Got %d, expected %d.", irq_fired, irq); \
+  } while (false)
+
+static status_t write_byte(const uint8_t addr[2], uint8_t byte) {
+  uint8_t data[3] = {addr[0], addr[1], byte};
+  return i2c_testutils_write(&i2c, kDeviceAddr, sizeof(data), data, false);
+}
+
+static status_t nak_irq(void) {
+  TRY(dif_i2c_irq_set_enabled(&i2c, kDifI2cIrqNak, kDifToggleEnabled));
+
+  // Write a byte to some random address.
+  const uint8_t kAddr[2] = {0x03, 0x21};
+  TRY(write_byte(kAddr, 0xAB));
+
+  irq_global_ctrl(false);
+  irq_fired = UINT32_MAX;
+  irq_global_ctrl(true);
+
+  // Wait for the write to finish.
+  TRY(i2c_testutils_read(&i2c, kDeviceAddr, 0, NULL, 1));
+  CHECK_IRQ_EQ(kDifI2cIrqNak);
```

Comment:
Is there any chance that the interrupt is fired between lines 111 and 113 which then gets wiped at 114? Would this cause any problems? Could we swap the order of the reset, or could other interrupts fire? e.g.

```c
  irq_global_ctrl(false);
  irq_fired = UINT32_MAX;
  irq_global_ctrl(true);
  
  // < another interrupt could fire here
  
  // Write a byte to some random address.
  const uint8_t kAddr[2] = {0x03, 0x21};
  TRY(write_byte(kAddr, 0xAB));
```

Reply:
All the interrupts are disabled except for the `kDifI2cIrqNak` (line 104), which is only triggered by the read operation (line 115), so I think the reset is done at the most appropriate time.

Reply:
Ah okay, not by the write. Thanks, that makes sense